parachain/src/wasm_api.rs

```diff
@@ -16,57 +16,50 @@
 
 //! Utilities for writing parachain WASM.
 
-use codec::{Encode, Decode};
-use super::{ValidationParams, ValidationResult, MessageRef, UpwardMessageRef};
+use crate::{TargetedMessage, UpwardMessage};
+use sp_runtime_interface::runtime_interface;
+#[cfg(feature = "std")]
+use sp_externalities::ExternalitiesExt;
+
+/// The parachain api for posting messages.
+#[runtime_interface]
+// Either activate on `std` to get access to the `HostFunctions` or when `wasm-api` is given and on
+// `no_std`.
+#[cfg(any(feature = "std", all(not(feature = "std"), feature = "wasm-api")))]
+pub trait Parachain {
+	/// Post a message to another parachain.
+	fn post_message(&mut self, msg: TargetedMessage) {
+		self.extension::<crate::wasm_executor::ParachainExt>()
+			.expect("No `ParachainExt` associated with the current context.")
+			.post_message(msg)
+			.expect("Failed to post message")
+	}
 
-mod ll {
```

Comment:
How does the new API look from a C standpoint? Previously, it was actually easy to write a parachain in C or C++. The new API is opaque and confusing - what is `self.extension` doing on the `Parachain` trait?

Reply:
That is all documented in Substrate: https://crates.parity.io/sp_runtime_interface/index.html